test/gpu/native/studies/sort/radixSortGpu.chpl

```diff
@@ -15,101 +16,198 @@ module RadixSort {
-      prefixSum[i] = sum;
+    for i in startIdx..endIdx {
+      var t : uint = arr[i];
+      arr[i] = sum;
       sum += t;
     }
   }
 
-  proc hillisSteeleScan(ref hostPrefixSum: [] uint, hostBuckets: int){
+  proc makePowerOf2(const sz){
+    // Powers of two only have the highest bit set.
+    // Power of two minus one will have all bits set except the highest.
+    // & those two together should give us 0;
+    // Ex 1000 & 0111 = 0000
+    if (sz & (sz - 1)) == 0 then
+      return sz;
+    // Not a power of two, so we pad it out
+    // To the next nearest power of two
+    const log_2_size = numBytes(uint)*8 - BitOps.clz(sz); // Calculates the log quickly
+    // Next highest nerest power of two is
+    const newSz = 1 << log_2_size;
+    return newSz;
+  }
+
+  proc hillisSteeleScan(ref hostArr: [] uint){
     // Hillis Steele Scan
     // This is the same as the count array, but each element is the sum of all previous elements
     // Uses a naive algorithm that does O(nlogn) work
     // Hillis and Steele (1986)
+    // if hostArr.size==0 then return;
+
     on here.gpus[useGpuId]{
-      var buckets = hostBuckets;
-      var prefixSum = hostPrefixSum;
+      // The algorithm only workd for arrays that are size of a power of two.
+      // In case it's not a power of two we pad it out with 0s
+      const size = makePowerOf2(hostArr.size);
+      var arr : [0..<size] uint;
+      arr[hostArr.domain] = hostArr;
+
       var offset = 1;
-      while offset <= buckets {
-          var prefixSumBuffer = prefixSum;
+      while offset < arr.size {
+          var arrBuffer = arr;
           @assertOnGpu
-          forall i in offset..<buckets { // foreach causes error? Investigate later
-            prefixSum[i] = prefixSumBuffer[i] + prefixSumBuffer[i-offset];
+          forall i in offset..<arr.size { // foreach causes error? Investigate later
+            arr[i] = arrBuffer[i] + arrBuffer[i-offset];
           }
           offset = offset << 1;
       }
 
-      // Change inclusive scan to exclusive
-      var prefixSumBuffer = prefixSum;
-      foreach i in 1..<buckets {
-        prefixSum[i] = prefixSumBuffer[i-1];
-      }
-      prefixSum[0] = 0;
-
       // This code is a more reader friendly way to do the same thing as above
-      // const log_2_buckets = numBytes(uint)*8 - BitOps.clz(buckets); // Calculates the log quickly
-      // for d in 1..log_2_buckets {
+      // const log_2_size = numBytes(uint)*8 - BitOps.clz(arr.size); // Calculates the log quickly
+      // for d in 1..<log_2_size {
       //   @assertOnGpu
-      //   foreach i in 0..<buckets {
+      //   foreach i in 0..<arr.size {
       //     if i >= 1 << d { // 2^d
-      //       prefixSum[i] = prefixSum[i-(1 << (d-1))] + prefixSum[i];
+      //       arr[i] = arr[i-(1 << (d-1))] + arr[i];
       //     }
       //   }
       // }
-      hostPrefixSum=prefixSum;
+
+      // Change inclusive scan to exclusive
+      var arrBuffer = arr;
```

Comment:
And again here?

Reply:
I guess we are shifting the array elements by one. We could do reverse iteration, but when you do `by -1` your loop becomes gpu-ineligible?

Reply:
Without the copy the code becomes racy. A reverse iteration would be nice but still racey, unless I'm mistaken

Reply:
Note to self: `foreach` loops are order-independent. :D